### ISSUE TITLE
New Functions and Requesting Community Input

### ISSUE THREAD

**Issue by kizniche** [kizniche/Mycodo#346]
The latest releases of Mycodo has a new category, Functions, with PID Controllers placed as a type of Function.

I'd like to get some input from the community as to what type of functions you think are useful or interesting that should be added.

Functions are, well, functions that carry out specific tasks. These are distinct from Math Controllers, which perform math on other controllers and store a new value (or values) in the measurement database. Functions typically do more than merely store a new value, such as with PID Controllers, which take an input and manipulate outputs in response to the input and PID settings.

Some new Functions I think would be useful:

1. [CoolBot](https://www.storeitcold.com/) clone
2. Air Conditioner/Heater (temperature sensor, heater output, cooler output) - This can currently be achieved with a PID controller, but more AC/Heater-specific options can be included to gear it more toward the specific use.
3. ???

**Comment by zsole2** [comment]
I always wanted to suggest it, now that you have asked for... irrigation control. With all the sensors, weather forecast checking, whatever. Like an opensprinkler clone,

**Comment by RynFlutsch** [comment]
Convert funktion:
e.g.:  Lux [lx]to [PAR](https://en.m.wikipedia.org/wiki/Photosynthetically_active_radiation) [µmol*m^−2s−1]
If there is a second sensor without raspberry functionality?

Get weather data online (maybe as Input)

**Comment by kizniche** [comment]
> I always wanted to suggest it, now that you have asked for... irrigation control. With all the sensors, weather forecast checking, whatever. Like an opensprinkler clone,

I'm not too familiar with the features of this, and I haven't found a comprehensive list of features on the website. Are there any specific features you have in mind? I'm sure Mycodo can already do a lot of what this can do, using just timers and conditionals alone, but I can see where there can be certain functions that utilize several sensors (such as float switch, temperature sensor, and rain sensor combinations) to determine if and how much a valve should be turned on.

> Convert function: e.g.: Lux [lx]to PAR [µmol*m^−2s−1]

What light sensor are you using? Would it be able to convert to PAR? Would a sensor [like this](https://www.licor.com/env/products/light/quantum.html) be more appropriate?

> If there is a second sensor without raspberry functionality?

What do you mean by this?

> Get weather data online (maybe as Input)

This is a good idea. I'll look into reliable weather sources with a good API.

**Comment by RynFlutsch** [comment]
> Would a sensor [like this](https://www.licor.com/env/products/light/quantum.html) be more appropriate?
Nice sensor.
Use Lux Sensor and [Quantummeter](https://www.specmeters.com/lightmeters/quantum-meters/#description) Kalibrate Sensor 1 with Sensor 2 
Yes supported Quantumsensor is the better way. Can be used as irrigation method.

**Comment by zsole2** [comment]
I'm not familiar either, just remembered it from ancient times, it was one of the first RasPi project. Just like beer-brewing, but I think Mycodo is already perfect for that without any further tweaking. After all, beer needs the  same as cheese, but at a higher temperature... (BTW, thanks for mentioning  CoolBot, I'll check it out, never heard of it).
With the garden irrigation, yes, I'd think mainly of duration timers. Then the timers should be able to be temporarily disabled, by e.g. soil moisture or an alternative sensor after rain. I did not think what conditional could be written, but I have not much experience with them. But this is a feature that should be included like a simple switch, If it rained, stop watering. The conditional for this should be built-in, if you look for user friendly additions.
Weather data  can also be integrated, if it is rain in the forecast, do not start watering (unless it does not come, then start after a preset period).
I did not think of the float switch, but it is indeed good idea when someone uses a cistern.
Maybe I'll find some time during the holidays to look into this.

**Comment by zsole2** [comment]
After getting familiarized myself with all the new features and changes in the interface, I'd have a few suggestions. 

Math controllers seem to be great and I remember old times when the seeds of those were there, but this is on a whole new level. On the other hand, these are just ways to manipulate inputs, aren't they? That is what you mentioned in the original post, setting them apart from Functions.
In that sense, they could be placed on the Input page, as a second addition line, and put into a second table, or even into in the same table as other inputs. Whoever uses this, will name them appropriately and know what they are doing, i.e. generating a new input value to be used later. 
In that sense, I think it would help if the input conditionals were in a separate table, just like for output. The user could quickly see what conditionals are set up, rather then remembering for what input they were set up by clicking the '+' signs.

In the same line of thought, timers are just a specific case for output manipulation, could also be merged.

That way the input page would have Add Input Device, Add Math Controller, and Add Conditional, while the output page would have Add Output Device, Add Timer, and Add Conditional.

With respect to Functions, it might worth some consideration whether the conditionals should belong here, based on your definition of a function: they perform specific tasks based on various inputs. I know this contradicts my previous paragraph, but just putting up ideas...

**Comment by kizniche** [comment]
I like compact interfaces, and I think Input and Math can be placed on the same page (along with Input and Math conditionals), keeping the menu name 'Input'.

However, I'm not sure I want to separate the conditionals from the actual controllers. Some people use a lot of conditionals (I see one install using 15 of them, with 8 activated, from the stats viewer). I think it would make it difficult to organize or understand which goes where if they're not listed under the controller they're being used with.

With Output and Timers, I think keeping them separate might be better, because some users have many outputs (I have 10, I see another install using 11), and timers (I have several, and see many installs using up to 8). If there were a low number being used, then it may be beneficial to have both on the same page, but for those who have a lot, I think it makes it easier to navigate and manager each when they're on separate pages.

As for conditionals being on the Function page, I think they would benefit from being associated with their respective controller. Although the person who sets up the system may know what's going on with them if they were separate, this may change if you come back after a long time and don't remember how you set it up, or bring a new user on to manage the system. If they are coupled to their controller, it is immediately obvious which controller the conditionals are acting upon, where as if they were in a separate list all together it would mean a lot of jumping around the interface to find which controller they belong to. This may become more confusing if the order of them isn't consistent, with several conditionals for one controller being out of order with other conditionals between them.

Thanks for the suggestion. I'll see about combining the Input and Math controllers to the same page, and think about your other suggestions. Mycodo is already pretty intimidating to new users and I want things to be as easy to understand as possible, even if it means keeping sharp distinctions between things like Outputs and Timers, or keeping groupings, like between a controller and its associated conditionals.

**Comment by zsole2** [comment]
I see your point on complexity and fully agree that it can be daunting for beginners ...but it is good to have choices, an I commend your great work, as I regularly  do...
As for jumping around when managing the system, I'd disagree. As for sensors and relays and timers (I also have a few of those, 4 sensors, 8 relays and 3 timers), those are 'set-up-and-forget-it' features for me: once added, they are there, and usually not visited again, unless some manual manipulation is necessary, These are relied upon in PIDs and conditionals, which provide all the output that is needed to me later on (graphs, notifications, etc.), and should be changed if something needs to be adjusted. Which is not really needed, one of my systems is still on v4.0.24 and up since about a year...
I'm not fully using conditionals (I have no relay conditional, and a few for sensors), but one of my issues with the current setup is when I do a fresh install, first come  the general config options, and then start to populate the parameters, go by the interface's logical order, so start with sensors. I cannot add the sensor conditional when I go along, because there are no relays  setup yet, so need to jump forward to the Outputs, then come back for the sensor conditional. If I could set up my sensors, relays and timers (the basics), then I would go to the 'effector' page, i.e.  Functions, and stay  there "forever".
Of course every user has different objectives to achieve, and ways to reach them. I also know that my case may not be the most generic. But  from a new user's standpoint, I think the basics should come easily... I'm not sure though what could be more important for a new user, everyone's might be different, for me, it was the PIDs, after setting up the sensors and the relays.

Just a crazy idea popped into my mind: would it be hard to get a 'Favorites' page? And letting the user put whatever he wishes there, whether it's a sensor, a relay, a conditional, a timer, whatever?

**Comment by kizniche** [comment]
The more I think about it, the more I like the idea of conditionals being separate from the controllers and placed on the Functions page with PIDs.

**Comment by kizniche** [comment]
> Just a crazy idea popped into my mind: would it be hard to get a 'Favorites' page? And letting the user put whatever he wishes there, whether it's a sensor, a relay, a conditional, a timer, whatever?

A favorites page could work but a lot of work would have to be done in preparation for such a thing, making all other parts more modular. I'm going to work first on moving Maths with Inputs and conditionals with PIDs, then think about the feasibility of a favorites page.

**Comment by zsole2** [comment]
Yes, I thought that it would be too much work to do favorites, and so was the idea deemed crazy. Probably there are more important things to do, anyway. Even if you got time, maybe you should do a poll to see how many people think it useful. Myself, I'm not even sure that it is... maybe only a 'nice-to-have' category... 
But if you achieve that kind of modularity, other thinks could become feasible, like a floating/popup window with a specific control, activated from e.g. the Live page or from the graphs...

**Comment by kizniche** [comment]
I had some inspiration and redesigned the conditionals to be more modular, and moved them all to the Function page. I also made a separate, dedicated Conditional controller that runs in the daemon, instead of the Math controller having its own conditional code and the Input controller having its own conditional code. Overall, things are much cleaner and work better. Thanks for the suggestion, it really paid off. Here's what the current iteration of the Function page looks like:

![screenshot-192 168 0 10-2017-12-22-17-46-01](https://user-images.githubusercontent.com/838427/34314307-435766ac-e740-11e7-8692-a3d27b56a98d.png)

**Comment by kizniche** [comment]
In doing this, PID outputs can now be used as Conditional inputs, so you can have an email notification if your PID isn't running (detected as no PID output), or other interesting things.

**Comment by kizniche** [comment]
I just finished testing all conditional typed (measurements, output, and edge detection), and am very close to making a release. I just have a few more things to test and tweak.

**Comment by Theoi-Meteoroi** [comment]
Awesome work.  I like the changes in 5.5 so far. I haven't been using the more advanced features quite yet as I'm focused on making sure them sensors be sensing.  

I would like to make a RFE - the ability to turn off the LCD backlight.  I want to use the assembly in a grow area with light and dark cycles. The dark cycles must not have light emitted from the LCD display during those periods.  Display updates can continue, however the backlight must be off.

**Comment by Theoi-Meteoroi** [comment]
BTW:  I like the PAR sensors y'all are talking about.  But way over my SRP for experimentation, I obtained this device - [https://www.hydrofarm.com/p/LGBQM](url)  It does have a serial port and a really crappy app to log the data.  I'll write a driver at some point soon that can at least put the data in Mycodo.  There are formula conversions between lumens and such, however light measurement is a complex topic and I doubt you could repurpose a general lux light sensor to umols/m-2/sec-1.  What you usually want is integrated PPDF - [https://www.apogeeinstruments.com/conversion-instantaneous-ppfd-to-integrated-ppfd/](url)

     Once you get to the actual mechanics of repeatable crop dynamics, you find out that Quantum PAR measurement and mols/day requirements are what drives modern productivity at scale.   The other thing I learned is that PAR is not what it used to be in respect to spectrum.  Most PAR charts have artificial roll-off in the blue and red in regions that have tremendous effect on phenotype expression.  The charts seem to reflect more of what we can produce artificially rather than what organisms can use photosynthetically.  This is an issue for sensors that are insensitive in those PAR roll-off areas, with the blue light being most energetic and typically unaccounted for in PPFD measurements.

**Comment by Theoi-Meteoroi** [comment]
I like the way the web modules keep running - I can use it to bounce the daemon while I iterate with my terrible coding skill.  But the road to Python3 seems like breakage ahead.  I've tried two upgrades from 5.4.15 and well, not good. We will probably have to exercise all the sensor code paths  to check for types issues. So I play this for the developer - [https://youtu.be/qjGSXGCi4Fc](url)  - Ima sure there are gonna be regrets for some.  I think this is a worthwhile turn - Python2 was too permissive with types IMHO.   And Apache2 was apparently not your friend, my second upgrade hangs at:

pi@Dev1:/var/log/mycodo $ tail -f mycodoupgrade.log 
[2017-12-25 23:53:27]   apache2-data apache2-utils libpython-all-dev python-all python-all-dev
[2017-12-25 23:53:27]   python-keyring python-keyrings.alt python-secretstorage python-wheel
[2017-12-25 23:53:27]   ssl-cert
[2017-12-25 23:53:27] Use 'apt autoremove' to remove them.
[2017-12-25 23:53:27] The following packages will be REMOVED:
[2017-12-25 23:53:27]   apache2
[2017-12-25 23:53:35] 0 upgraded, 0 newly installed, 1 to remove and 4 not upgraded.
[2017-12-25 23:53:35] After this operation, 591 kB disk space will be freed.
(Reading database ... 138637 files and directories currently installed.)
[2017-12-25 23:53:37] Removing apache2 (2.4.25-3+deb9u3) ...

Yup. That timestamp was ... about 17 minutes ago. Its dead, Jim. The web upgrade status hung even earlier at:

[2017-12-25 23:50:08] 
[2017-12-25 23:50:08] 
[2017-12-25 23:50:08] Installed version: 5.4.15
[2017-12-25 23:50:08] Latest version: 5.5.1
[2017-12-25 23:50:08] 
[2017-12-25 23:50:08] #### Upgrade to v5.5.1 initiated 2017-12-25_23-50-04 ####
[2017-12-25 23:50:08] 
[2017-12-25 23:50:08] #### Beginning Upgrade: Stage 1 of 2 ####
[2017-12-25 23:50:12] Stopping the Mycodo daemon...Done.
[2017-12-25 23:50:15] Downloading latest Mycodo version to /home/pi/mycodo-5.5.1.tar.gz...Done.
[2017-12-25 23:50:15] Creating /tmp/Mycodo-5.5.1...Done.
[2017-12-25 23:50:15] Extracting /home/pi/mycodo-5.5.1.tar.gz to /tmp/Mycodo-5.5.1...Done.
[2017-12-25 23:50:15] Removing /home/pi/mycodo-5.5.1.tar.gz...Done.
[2017-12-25 23:50:15] Copying /tmp/Mycodo-5.5.1/.upgrade status file to /tmp/Mycodo-5.5.1...Done.


I don't have a bunch of data I can't live without so I typically re-install when upgrade fails.  I did have a successful upgrade from 5.3.0 once, it was a beauty.   I'll try to be more helpful with post 5.5 upgrades however I didn't want to get stuck at a down-rev state.

**Comment by leoneo** [comment]
Hi,

I have a proposal for a new function.

I would like to install several RPi systems with mycodo (hive, henhouse, weather station, greenhouse, fish tank,...) and have them all on the same manager, with the possibility of having interactions between them.

Thank you

**Comment by RynFlutsch** [comment]
Timespan Timer: Start and Endtime with Seconds

Conditional: If Output x turns on/off for x seconds or any duration
then turn on/off output y after time x

**Comment by zsole2** [comment]
New function: ability to get a quick overview of Mycodo through terminal. Imagine the Live page, all important data, but on the command line. Could also include system info, too.

**Comment by kizniche** [comment]
Good suggestions. I'll think about how to implement a few of these. Regarding the command line monitoring, I have been wanting to do this for a while (#17, so you can see how old that issue is).

**Comment by zsole2** [comment]
Ah, those good old days! Speaking of those open issues, what about #45? I remember seeing a request for something similar recently...

**Comment by kizniche** [comment]
I have a rudimentary system working called Remote Admin (accessed through the Config menu). Right now you can only add other Mycodo systems (authenticate) and it will compile and show all measurements from all added systems. I only recently got that all working, so right now it's very basic and mostly a proof of concept. I'd like to turn it into a full-fledged remote admin console, but it's pretty overwhelming to think about what needs to be done to get to that point. I have a constant pull in different directions to develop but I try to work on stuff that interests me at a particular time in order to keep my work on this fun instead of feeling like true work.

**Comment by zsole2** [comment]
Agree completely! Keep it fun if you do it for fun and do what interests you. I think all Mycodo users should be  eternally thankful for your exceptionally fast reaction on any issue that is raised. 

So, does putting a relay control on the Dashboard sound fun? Like a series of red/green dots in a box, clicking any of those changes the relay's status. It could be quite useful. Similarly, temporarily disabling sensors and PIDs could also be useful. 

In this respect, I always wanted to request this: if a sensor is turned off, there is a notification on the  related PID being turned off. Can it be done the other way, too? If I turn on a sensor, Mycodo could ask  me if I want to turn on the PID. This could save a few page turns.

**Comment by kizniche** [comment]
Now that the Graph page has turned into a true Dashboard, those sound like really good elements to add to it. I'm not sure how the multiple-activation element would fit in, other than it can be a function (activate/deactivate multiple function) and then that can be added to the Dash.

**Comment by zsole2** [comment]
Speaking of new features: export/import for settings.
I mean not the saving of the full database, but each parameters separately. Those could be imported into a new install, even if the database version is already ahead a few cycles. I know this could be a lot of coding, and checks everywhere, but having e.g. a text/tagged file with all the parameters and a possibility to import would allow changing quite a few things at the same time, and reload once. This could result a lot of screwed up systems, so you'd need a disclaimer that it is  for advanced users, and no support, go for reinstall.

**Comment by Theoi-Meteoroi** [comment]
I agree with zsole2, keep it fun fer yerself.  I can see that parameter export/import would hinge on database consistency/stability and strict versioning.   And the pain from that lasts ( in code overhead ) until you break/renew the version bond at major releases.   At this point myself, I'm not hung up on database version skew.  When I deploy, the versions are frozen (unless BUG!) until the task is complete for whatever task it was assigned. I treat it as an appliance at that point - compatible with similar/same versions, or until Kyle decides otherwise. :)

**Comment by RynFlutsch** [comment]
> Conditional: 
If Output 1 turns on/off 
for x seconds or any duration
then turn on/off output 2
after time x [s]

I would like to controll a bigger amount of electrical loads with high Inrush Current.
So the goal is to activate the relays one by one with some seconds delay and turn them all together off.

I think i am the guy who u saw with so much timers ^^ (around 20)
It realy suxx now if i want to change the times.

Example:
Timer Output 1 on 07:00 - 16:00
off 16:01 - 23:00
on 23:01 - 01:00
off 01:01 - 6:59

(4 timers needed for 1 Output)

Conditional for Output 2 -5
behave exactly like the timers for Output 1 but with 5 seconds delay for each output @ turning-on.

Or maybe a dublicate function? So i have only to change the gpio pin.

**Comment by kizniche** [comment]
I could add a delay option to the Output Conditional Actions.

> Timer Output 1 on 07:00 - 16:00
> off 16:01 - 23:00
> on 23:01 - 01:00
> off 01:01 - 6:59

Also, an alternative way to setup this (theoretically), is to create PWM Outputs, then create a PWM Method and set when you want your relay to be on to 100 (100% duty cycle) and when you want it to be off to 0 (0% duty cycle). This will act essentially as an on/off instead of an actual PWM, and you can have it all work form one timer. You can also set up different methods so you can easily switch between timer regimens with just a few clicks.

**Comment by zsole2** [comment]
I have no experience with high power devices on the relays, but on my PiZero the timers starting in sequence has quite a good delay:
```
2018-01-31 06:46:01,439 - mycodo.timer_1 - INFO - Activated in 550.9 ms
2018-01-31 06:46:03,494 - mycodo.timer_2 - INFO - Activated in 1555.1 ms
2018-01-31 06:46:05,701 - mycodo.timer_3 - INFO - Activated in 1479.9 ms
```
Finally a situation where it is good to have a slow system.
Checking back historically, starting the timers takes around 500-1500 msec, while stopping is in the range of 5000-7000ms.
Isn't this amount of delay enough for you?
And  if there is an inherent  delay anyway, @kizniche can include a little more, for general usage a few secs shouldn't matter in these daily timers.

**Comment by pekkakorre** [comment]
Hi,
Thanks for this awesome system!

Possibility to make notes would be nice, kinda diary. Maybe just simple where to write things down, or tap point of graph example strange peak in graphs can be explained.. Maybe timed reminders "fill humidifier!".

I dont know is it possible to make Output Function to start taking timelapse? It would be nice too.

**Comment by kizniche** [comment]
There used to be a notes feature that also showed the notes on a graph, in a previous version. I should add that feature again.

And I also like the idea of a conditional action triggering a timelapse start.

**Comment by RynFlutsch** [comment]
Hello thanks for the input, was quite busy ;)

> I could add a delay option to the Output Conditional Actions.
The Outputs all turn off, at the same time ?

> Alternatively, I could create a timer similar to the PWM Method timer, that would be a Relay Method timer, which would accept a Relay Output and an On/Off Method.

Sounds great.
Possible to extend that Method-range over Days / Weeks / Months ;) ?

So users have the possibility to build their own day. i.E. with 22h.

**Comment by frodus17** [comment]
It would be cool to be able to move around the dashboard items. Also add things like indicator lights to show status. The bar graphs work well to show on/off currently.

**Comment by RPi-Growbox** [comment]
I used this script to make Output conditional timelapse, maybe it can be used:
`#!/bin/bash
# Timelapse controller for USB webcam

DIR=/home/pi//Mycodo/timelapse

x=1
while [ $x -le 1440 ]; do

filename=$(date -u +"%d%m%Y_%H%M-%S").jpg

fswebcam -d /dev/video0 -r 1280x720 $DIR/$filename

x=$(( $x + 1 ))

sleep 10;

done;
`

From: [https://www.makeuseof.com/tag/how-to-capture-time-lapse-photography-with-your-raspberry-pi-and-dslr-or-usb-webcam/](url)


> It would be cool to be able to move around the dashboard items. Also add things like indicator lights to show status. The bar graphs work well to show on/off currently.

You can move dashboard items with those Up/Down-buttons. Also like idea of indicator light.

**Comment by frodus17** [comment]
RPi-Growbox - Yeah, I'm imagining something like being able to click and drag around. Up and down is OK. I just started playing with MyCodo this last Saturday.

**Comment by kizniche** [comment]
> Indicator lights to show status of outputs
> Digital display to show at a glance the temperature, humidity, etc.
> Buttons or toggles to turn outputs on/off either momentarily or toggle
> Input for a setpoint for the PID

All great ideas. The outputs on the dash has been suggested before and I had it on my mind to add. The large digital display would be a nice addition. I'll also think about how to add the Input as a PID Setpoint... Can you give me an example of a use for this (I'm having difficulty seeing utility with this one).

**Comment by frodus17** [comment]
Concerning adding an input for PID setpoint: Homebrewing, where I need to set a temperature easily. I'd have the Hot Liquor tank and Boil kettle. I can set PID's up for the HLT and Boil kettle. I can still go into PID's and change there, just thought it'd be nice to set temperature from your dash. 

I can do everything I need now. I just saw the thread where you were looking at ideas.

Another would be Sous Vide, just have it all on one screen. The set up is all done on other pages, so just doing small things like Setpoint on the dash would be cool.

**Comment by zsole2** [comment]
In the spirit of the top of the issue, where I believe generalized Functions were sought (but I like that all kind of improvement ideas are welcomed), I also join to propose a specialized  **Sous-vide** controller. 
In fact, I just built one over the weekend from leftover components, and seems to work well. I will summarize my thoughts on this soon, whenever I have time to collect them and write it down, as well as more experience is gathered.

**Comment by zsole2** [comment]
Onto  the wishlist of minor  of improvements:  for PID tuning purposes it would be useful to see the Kp, Ki and Kd values on the graph.

**Comment by zsole2** [comment]
For the Sous-vide controller, or other major appliance style changes (like Colbot, AC control, irrigation control), the interface should somewhat redesigned, to allow for a specialized start screen for the specific appliance. 
It should include the ability to set the startpage Mycodo comes up with, and probably locking out users from the other pages (thinking of my wife with the sous-vide, she would be frightened from the myriads of options). It can be debated whether this specialized page should be designed specifically with its own controls, or the Dashboard could be adapted. 
For the sous-vide, ability to include simple touch-button interface would be great plus. Adafruit has a combo LCD-keypad modul with I2c interface, that could be easy to implement. Although I'd prefer something more granular, with ability to leave the buttons detached  from the control PCB, but this would obviously include more coding and planning for what the buttons should be able to operate within the whole Mycodo interface. 
On-off button(s) are now  implemented outside of Mycodo, but what if they could be easily included as an Input?
OK, now my commute soon ends, so leave it at this...

**Comment by kizniche** [comment]
I just finished adding a new Dashboard element for monitoring and controlling Outputs. I think it came out quite nice. Suggestions for improvement are welcome.

![screenshot-10 48 44 226-2018-02-16-19-12-38](https://user-images.githubusercontent.com/838427/36335387-90edf9ba-134d-11e8-9eaa-d9753bce96eb.png)

**Comment by zsole2** [comment]
This is great progress!!
If I'm thinking about adding my 8 relays, this would be a little oversized, though. I'd prefer a smaller footprint, with less info at first (simple on/off buttons, last value), and after extending it you can get all other data and the settings, of course.
It never occurred to me until now, is there a specific reason for having 1/12th step for the with of the elements?

**Comment by kizniche** [comment]
> It never occurred to me until now, is there a specific reason for having 1/12th step for the with of the elements?

On very large screens, 1/12 can be a lot of space. It's in 12th increments because that's the column system used by Bootstrap, the framework I use for the UI.

Here's after a bit more tweaking, showing the variability of how the elements can be set up. If the Output Controls aren't featured, they will now appear in the dropdown.

![screenshot-192 168 0 10-2018-02-17-16-45-35](https://user-images.githubusercontent.com/838427/36345921-6d16b224-1402-11e8-89a8-631975bf34f6.png)

![screenshot-192 168 0 10-2018-02-17-16-46-21](https://user-images.githubusercontent.com/838427/36345922-704735f4-1402-11e8-9573-8274e6c81264.png)

**Comment by kizniche** [comment]
Since the Output  Dashboard element is so similar to a Measurement element, I was able to easily add that with a little modification to the Output element. Configuration is similar to the Output:

![screenshot-192 168 0 10-2018-02-17-18-04-18](https://user-images.githubusercontent.com/838427/36346472-3cf07624-140d-11e8-9176-44e54b275da4.png)

**Comment by kizniche** [comment]
I think allowing the number of decimal places to be defined would be a good improvement.

**Comment by kizniche** [comment]
I just added the ability to view the P, I, and D values of a PID controller on graphs. This should make it easier to tune. This was a great suggestion.

**Comment by kizniche** [comment]
> the interface should somewhat redesigned, to allow for a specialized start screen for the specific appliance.

Like a Kiosk mode?

> For the sous-vide, ability to include simple touch-button interface would be great plus.

I have a 7-inch touch screen (1024x600) on one of mine. It auto-logs in when booted up and automatically starts the web browser and full-screens it.

https://www.waveshare.com/product/7inch-hdmi-lcd-c.htm
https://www.amazon.com/Waveshare-Capacitive-Interface-Raspberry-Beaglebone/dp/B015E8EDYQ

**Comment by kizniche** [comment]
I've been working on adding a PID Control element to the dashboard. This would allow viewing all the parameters of the PID, including P, I, and D values as well as control the PID. This would be handy for making an interface for a particular appliance, such that your wife you mentioned could feel comfortable using.

**Comment by zsole2** [comment]
> On very large screens, 1/12 can be a lot of space. It's in 12th increments because that's the column system used by Bootstrap, the framework I use for the UI.

I was more like going for 1/16th... But if that is the limitation of the programming, then 1/12 is OK.

> Here's after a bit more tweaking, showing the variability of how the elements can be set up. If the Output Controls aren't featured, they will now appear in the dropdown.

What I think would be useful to have at least an  On/Off button visible to be able to quickly control the output.BTW, can it be a single button? On-button if the output is turned off, and vice versa. 

> Since the Output Dashboard element is so similar to a Measurement element, I was able to easily add that with a little modification to the Output element. 

Very nice! Now that I spent a few hours to perfect my gauges, I'll have to decide to use this instead.

> I've been working on adding a PID Control element to the dashboard.

Can't wait to try it!

> Like a Kiosk mode?

Yes, that is  the word I was looking for.

> I have a 7-inch touch screen (1024x600) on one of mine. It auto-logs in when booted up and automatically starts the web browser and full-screens it.

I remember that you have mentioned it somewhere here, but it is too much for  simple control. And for Sous vide, it can be quite splashy...

**Comment by kizniche** [comment]
I just finished the first iteration of the PID control element. It came out pretty nice. I integrated the Output green/red to indicate if it's on or off and the refresh-less button press that activates/deactivates it, like the Outputs have.

![screenshot-192 168 0 10-2018-02-19-22-08-35](https://user-images.githubusercontent.com/838427/36406001-c86cc36e-15c1-11e8-98ee-3790aa3d11a9.png)

**Comment by kizniche** [comment]
I'm probably going to be tweaking the display of these. I want to add units to the PID setpoint and measurement values. I also want to add a checkbox to enable the display of more detailed information like P, I, and D values, output value, etc.

**Comment by kizniche** [comment]
I think I'm going to make a release so people can start using it and give feedback about how it can be improved.

**Comment by kizniche** [comment]
It works pretty nice alongside the graph, for easy viewing of current conditions.

![screenshot-192 168 0 14-2018-02-19-22-33-23](https://user-images.githubusercontent.com/838427/36406469-13c93150-15c5-11e8-9daf-988b130cf78a.png)

**Comment by zsole2** [comment]
> ... and the refresh-less button press that activates/deactivates it, like the Outputs have.

That will be very useful! It would also be good to implement throughout, not just here and for the Outputs. And I just realized a few days ago that it is  possible to change the PID setpoint without deactivating the controller. It would a useful feature at  more places, LCD messages come to my mind... or changing Input periods.

**Comment by kizniche** [comment]
The issue is WTForms does not play nicely with the ajax form submission (refresh-less form submission). I got it to work for buttons that have a dedicated task, but refactoring the entire form system of Mycodo would be an immense feat.

**Comment by kizniche** [comment]
The P, I, and D values added to the dash element really helps with understanding what's going on with tuning. Coupled with showing them on the graph should help a lot with tuning.

![screenshot-10 48 44 226-2018-02-20-15-59-06](https://user-images.githubusercontent.com/838427/36448925-280da2a8-1657-11e8-9ede-c9e69bcad503.png)

**Comment by zsole2** [comment]
Someone above mentioned reordering elements by drag-and-drop. I guess is not feasible without a lot of work. 
However, having a dozen items makes using up/down quite time consuming to get the desired placement. Like now, for testing the new dashboard items, and moving them from 14th place to the top to see them easily...
I can imagine that setting a position number for a given element and then move it into that position could be much easier to code than implementing drag-and-drop, but get us much batter control.

**Comment by frodus17** [comment]
I think that'd be the easiest, rather than having to click up a dozen times.

Also, is there a way to have a master page save? rather than in each page, work on a few things on a few elements and have to hit save on each one at a time? I forget sometimes that I changed something on another element, and have to go back through it again.

**Comment by zsole2** [comment]
The dashboard widgets for Output and PID reset the Refresh period to None when a setting is changed. If I remember correctly, this was the situation with the gauges, so should be an easy fix.

**Comment by zsole2** [comment]
My point with reordering of the elements of the Dashboard (and basically with everything these days as Mycodo gets more and more feature-rich)  is to avoid long waiting times. With the Pi Zero, redrawing all the graphs could be painfully slow. 
Another idea would be to hide the elements. A button for 'Show/hide graph' could be there next to the current 'Show/hide', which is for the graph's series. This seems to be easy to implement with highchart.
Implementing this for Inputs  would in practice could eventually replace the Live page, if someone sets up all his controls.

**Comment by kizniche** [comment]
> The new dashboard widgets (similarly for Input, Output and PID) reset the Refresh period to None when a setting is changed.

I noticed this and pushed a fix for this yesterday.

Reordering the elements is tricky. I think it would take a lot of effort to implement a drag-and-drop system. I will look into whether there are pre-built systems that would integrate well with Mycodo.

**Comment by zsole2** [comment]
I have just experienced the the graph to reset the Refresh period, too.

**Comment by zsole2** [comment]
I realized that I will definitely need hardware buttons for the sous-wide device to turn on/off/pause  the PID while opening the lid, due to  the sensor arrangement. Fortunately I remembered the good old days with terminal commands for the client and daemon, so i tried `mycodo-client  --deactivatecontroller PID 01`. Got this error: `Server returned: (1, 'PID controller with ID 01 not found')`. So how do I find the ID of a controller?

**Comment by kizniche** [comment]
The unique ID is a long alphanumeric string. It can be found when you hover your mouse over the numeric PID ID on the Function page.

**Comment by zsole2** [comment]
The cursor changes when I hover there to a crossed-out-circle, and no number is shown.

**Comment by kizniche** [comment]
If you have ```Hide Form Tooltips``` checked in the General Settings, it won't show the tooltip with the unique ID.

![screenshot - mycodo pid unique id](https://user-images.githubusercontent.com/838427/36746866-629e53ee-1bc1-11e8-94d5-72d831f4777f.png)

**Comment by kizniche** [comment]
You can also look at the source of the page, or right-click the disabled input and select ```Inspect``` to see the ID:

![screenshot - mycodo pid unique id 02](https://user-images.githubusercontent.com/838427/36746960-a14bc07c-1bc1-11e8-81c6-c9b84d342cc8.png)

**Comment by kizniche** [comment]
> I realized that I will definitely need hardware buttons for the sous-wide device to turn on/off/pause the PID while opening the lid

You can use switches that are monitored by an Edge Detect Input. This would allow you to turn the PID on/off, though it won't allow you to pause/resume it. I could add pause/resume as a conditional action to make that possible. Then you can have a switch for on/off, and a switch for pause/resume.

**Comment by zsole2** [comment]
Thanks! I'll play with the Edge Input. Or maybe two GPIO State conditionals with a double-pole switch.

**Comment by kizniche** [comment]
It was easy to add, so I just went ahead and did it ;)

If you want to upgrade to the latest github commit (ahead of the release), all you need to do it edit Mycodo/mycodo/config.py and change at the top ```FORCE_UPGRADE_MASTER = False``` to ```True```, then restart the web UI from [Gear Icon] -> Restart Frontend, then go to the Upgrade page and you'll be able to upgrade to the git repo master.

**Comment by zsole2** [comment]
Started to set up my switches and conditionals. There are a few problems.
First, the Edge conditional cannot read the state of the pin, giving the error: 
```
2018-02-28 13:04:51,459 - mycodo.conditional_1 - ERROR - Exception reading the GPIO pin
2018-02-28 13:04:51,461 - mycodo.conditional_1 - ERROR - GPIO not configured correctly or GPIO state not verified
```
This was the setting for the conditionals:
![image](https://user-images.githubusercontent.com/11209960/36787201-8d6d541c-1c89-11e8-862d-31a06fa6b029.png)

I was just to curious to see what happens, because there was some strange behavior with the Edge detection. Or the conditionals, I may not know. Sometimes, the switch registered the activation/deactivation of the PID in the log, sometimes not. Even when it was not logged, the action may have been carried out, since re-triggering the action through the GUI give the warning that the PID is already at the desired state. 
With the pause/resume, the same happened, and even registered a paused state in the log when the controller seemed to be deactivated.
Unfortunately, the GUI is not updated according to the conditional action, neither in the Functions page, nor the Dashboard. The color of the controls and the buttons do not change.
So, there are a lot of issues here, how should  I proceed?

**Comment by kizniche** [comment]
Let's start new issues in their own issue, and keep only discussion of new features.

**Comment by zsole2** [comment]
OK, can you move it or should I refile it?

**Comment by kizniche** [comment]
Just delete the comment here and create a new one. I don't have any ability to move it as the owner of the repo. I'm on the road to a field site I'm working on today, so I can't do anything except short replies on my phone.

**Comment by zsole2** [comment]
Feature request: ability to hide dashboard elements.
I think this would be a useful addition. 
Same as the ability to somehow improve reordering, as discussed somewhere earlier...

**Comment by kizniche** [comment]
I thought of a way that makes organizing dashboard elements simple and mobile-friendly.

![screenshot-192 168 0 10-2018-03-23-00-46-33](https://user-images.githubusercontent.com/838427/37812036-1ed181ba-2e34-11e8-95c8-8fceafc3b64e.png)

**Comment by frodus17** [comment]
I like that. Also, could you hide the Select a dashboard object? It sort of sticks out. Maybe a grey "lock" button to lock and unlock. Locking would get rid of the "+", and freeze everything into it's spot. Unlock would be the view we currently have, with the ability to add/edit dashboard elements. 

Any chance we could get °F to display in parenthesis on anything with a temperature reading?

**Comment by zsole2** [comment]
I really like this!
Instead of hiding "Select Dashboard Object" button and the selection choices, all could go into the menu, just below Dashboard, titled e.g. "Edit Dashboard elements" to open up a new page. This would make Dashboard really clean...
I have no problem with the "+", it is good to check the current settings.

**Comment by kizniche** [comment]
As a middle ground, I made the Add/Reorder Element options smaller to be less obtrusive. I added the output/PID popups to be controlled by the "Info", "Success", and "Warning" popup suppression options in the General settings.

As for Fahrenheit conversions from Celsius, I've done this many places. Where is it specifically not performed so I can look into adding it?

**Comment by frodus17** [comment]
Reorder is nice, looks slick. 

For Fahrenheit conversions, I don't see that any of the Dashboard elements support Degrees F. The graph, measurement and gauges don't show an alternative F.

Would it make more sense to have 2 different sensor types for temp sensors? I'm not sure where it makes the most sense to have a change. I would think an option in settings to have Imperial versus Metric. I'm just used to working in all degrees F.

**Comment by frodus17** [comment]
Kyle,
Graph element: Is there a way to change what the Y-axis displays for units?

Gauge element: I cannot change the main display value of the gauges, and I've got to hover over it. 

Measurement element: I don't see anything else when hovering over this element. It seems to only ever show temperature in C.

I'm using this for an actual dashboard for a kegerator and fermentation chamber with a touchscreen. Right now, I've got to touch each dashboard element.

**Comment by 1derjoints** [comment]
Couple people I know are wondering the same.

Can we make the main display values of the gauges display in default in Fahrenheit instead of celsius?

**Comment by Magnum-Pl** [comment]
I'd like to be able to display Fahrenheit in the gauges as well. With my setup, it would be very convenient to be able to glance at the display instead of having to use an input device to hover over the gauge to get the Fahrenheit value. 

What I'd really like to see is another gauge type that shows only large numbers for the value. The option to choose the color of the numbers to differentiate one gauge from another at a glance, as well as the option to have the colors change throughout the range of the gauge would be fantastic.

**Comment by kizniche** [comment]
I've been thinking about it for a while and I think I've come to the conclusion the best way to approach this is to do the conversion before being stored in the measurement database. I was just able to implement the ability to store any temperature measurement as C, F, or K. I'll be able to add more potential conversions for other measurements and allow user-created conversions between two measurements.

**Comment by kizniche** [comment]
I still need to work on the graph page, but so far it turned out to be a relatively easy conversion system to get working.

![screenshot-192 168 0 10-2018-04-01-20-23-44](https://user-images.githubusercontent.com/838427/38178950-f064d246-35ea-11e8-9781-7da35bccc7ce.png)

**Comment by kizniche** [comment]
![screenshot-192 168 0 10-2018-04-01-22-23-00](https://user-images.githubusercontent.com/838427/38180548-6b203c40-35fb-11e8-9ab9-9b8a07d7bdd5.png)

**Comment by frodus17** [comment]
Kyle,
I really like how this is implemented. Thanks!

**Comment by kizniche** [comment]
+1 feature, +20 bugs

This has been a challenge

**Comment by frodus17** [comment]
Sending some money for beer :)

**Comment by kizniche** [comment]
Thanks! I think I've finally ironed out most bugs. We'll see upon release day.

**Comment by frodus17** [comment]
Kyle, looks great! I had to delete my sensor and re-add it. It didn't like being changed. 

Live measurements look good. Dashboard looks good. Measurement, Graph, Gauge and PID all work and display the right temperature. PID changes to not store as negative looks to be working. 

I'll test sunrise/sunset overnight with an output/graph.

Thanks again for doing all this!

**Comment by ralphknoops** [comment]
I would love to see some weight sensors added to Mycodo for tracking the weight loss of my aged meat.

My Scenario would be:
When weight of the meat is 20% less than the start weight I would like to receive an alert.

If I have the weight measurement I can realize the above scenario with a Conditional quite easily.

I'm doing some testing at the moment with a load cell and a hx711 with a simple script i found.
https://github.com/tatobari/hx711py

- You need to fill in twee GPIO pins 
- Start a one time calibration (setting the reference unit). 
- Ready to start :)

**Comment by kizniche** [comment]
Add an Equation Math controller and merely put your desired value in the equation field, then activate. This creates a flat line at that value.

**Comment by kizniche** [comment]
@ralphknoops How has the hx711 been working?

**Comment by ralphknoops** [comment]
Not so great yet. 
I made a test setup: [https://farm1.staticflickr.com/949/41079098194_9557f041e5_b.jpg].

But...I'm having problems getting different readings from the HX711 every time I restart the script. I've got two load cells and two HX711 and both with the same problem so definitely a problem.
Because i'm not that familiar with python I'm going to check with a friend of mine this Thursday to try and fix this. 
I'm using this script https://github.com/tatobari/hx711py. 
I'll do an update by the end of the week, hope it works :).

**Comment by kizniche** [comment]
Have you tried changing these settings?

https://github.com/tatobari/hx711py/blob/ada5bee41056a51b204a1dfca3aa7fee4b3963d3/example.py#L14L21

**Comment by kizniche** [comment]
@ralphknoops Have you figured out your weight sensor issue?

**Comment by ralphknoops** [comment]
@kizniche  No, had priority on other things last month so didn't had time for this. I see you made some new functionality available which i can use for the weight sensors.
I'll try to fix the issue somewhere this or next week, and upgrade Mycodo :)

**Comment by Theoi-Meteoroi** [comment]
In the last couple of weeks, Adafruit has been deprecating some python sensor libraries used by Mycodo ( and thousands of other projects ) and archiving them in a state that seems to have broken the pip install process. The carefully crafted work that Kyle performed to keep these as well supported as possible in the project as options was mostly undone.

I can see that the pip developers have also recently documented a deprecation process that factors into this problem with the repositiories. 
[https://pip.pypa.io/en/stable/news/](url)

These are the ones that have been archived in the last 16 or so days:
https://github.com/adafruit/Adafruit_CCS811_python
https://github.com/adafruit/Adafruit_AMG88xx_python
https://github.com/adafruit/Adafruit_Python_BMP
https://github.com/adafruit/Adafruit_Python_TMP
https://github.com/adafruit/Adafruit_Python_MCP4725
https://github.com/adafruit/Adafruit_Python_MCP9808
https://github.com/adafruit/Adafruit_Python_BME280

I have pulled the archives of a few of these and manually installed them.  Next running the dependency tool in v6.2.2 and restarting the mycodo backend and then rechecking to see if the dependency was satisfied seems to install them in Mycodo.

However this is a problem. These projects cannot accept issues ( read-only ) and will never be updated from their current state. No bugfixes.  No improvements.   We apparently need to create community libraries that will not be subject to arbitrary business plans and policies.  These modules represent a significant percentage of the most common low cost, high precision measurement devices used in open-source projects.  I am quite disappointed with how Adafruit implemented their End of Life policy in the case of these libraries.

**Comment by kizniche** [comment]
For the time being, I've indicated a version number that directly precedes the deprecation version so that new installs won't error when trying to install these libraries. I just pushed this release so to avoid issues with new users. I haven't really had the time to think too much about how to move from here, but at last for the time being, I think my fix should work.

**Comment by Theoi-Meteoroi** [comment]
It seems Kyle has already fixed the dependency issue by selecting previous version with the pip install.

Thanks Kyle!

**Comment by kizniche** [comment]
I guess if any issue comes up where one of these libraries needs to be edited, I can edit my forked repositories (when I backed up all of these in anticipation of Adafruit doing something shady) and modify the Mycodo code to install from my updated repositories.

**Comment by Theoi-Meteoroi** [comment]
I've been working with a breakout that includes a Si7021, a BMP280 and the CCS811.  The Adafruit library calls out a device ID issue that I'm debugging now. The breakout is a CJMCU-8128.  I have found a couple of interfacing issues that might require mitigation to make this device work reliably.  One is a clock-stretch requirement when a measurement is requested. The other is the possibility of having downrev firmware on the embedded microcontroller internal to the CCS811. There are a couple of other pins besides I2C and power that are really needed to make the device behave properly according to the vendor documentation.  The simple way to interface to the device seems to be through a CP2112 USB to I2C bridge like the vendor's evaluation board.

**Comment by Theoi-Meteoroi** [comment]
For your consideration:
I think I mentioned a particle / dust detector I was looking at that is made by the same folks that make the MH-Z CO2 sensors.  The ZH03B starts up in a mode that streams binary over serial @9600 baud.  I use a USB to UART bridge to interface the device to a Pi.  I bought them while we were having the fires here last October because we were sucking harmful smoke and soot with every breath. There was no way to figure out if the indoor air was any better.

  I tested the 2 units I have by comparing readings, introduce a smoke source then clean the air with HEPA filtration.   I've included in the wiki a research document that compares this sensor with several others of similar type and reviews internal design features.

 It has taken me quite awhile to get around to writing a Python driver to snag the data it provides in python3.  Python2 was easy but not really robust. And these days I want Python3 to be Mycodo compatible.

I've just made the repo public after testing the code and creating some documentation.

[https://github.com/Theoi-Meteoroi/Winsen_ZH03B](url)

The code currently takes one sample and displays on stdout:

PM1 level:  42 PM2.5 level:  79 PM10 level:  93

You are welcome to just add the code to Mycodo base. This would extend the environmental sensing footprint in air quality monitoring.    The vendor is pretty responsive when contacted.

**Comment by 9600** [comment]
You called?

**Comment by Theoi-Meteoroi** [comment]
Ah, sorry about that.   Not intentionally.

**Comment by kizniche** [comment]
@Theoi-Meteoroi That's great! Of course I'll add it. Thanks for all the great work.

**Comment by Theoi-Meteoroi** [comment]
Yowsa.  Pushed that upgrade button to get 6.2.4. ( IMHO really worth the effort you put in to make it work).

Lets take a look at that Dashboard:

![chart 3](https://user-images.githubusercontent.com/16455466/45005052-1d8cd180-afa6-11e8-8d15-e231220226e7.jpeg)

You can see where I lit up a cigar ( for science ) and produced some data.  The sensor is 3 feet from the front door and about 24 inches above a HEPA room filter.    I've watched that HEPA filter bring the counts down to the teens on cli.  THIS is way more fun and saves me from having to figure out Grafana so I can focus on further development of the driver.

The driver is really basic, it just listens and lets the little fan run continuously. That costs about 120mA in power. Sleep mode is ~10mA.   There are environmental scenarios where you might want to sample occasionally or perhaps stop sampling because conditions may damage or prematurely age the device.  Deploying as a handheld instrument is one. Condensing humidity conditions could allow particles to adhere to the internal surfaces.   I see more work could be done in that direction.

Thanks for adding this Kyle!

**Comment by kizniche** [comment]
That looks great! It's also nice I was able to get the sensor working with the first try adding it- developing the new unit/measurement system really payed off. There are definitely some more options I'd like to add for different applications.

**Comment by Theoi-Meteoroi** [comment]
I took a look at the code - it looks awesome!  I can focus on solving point problems with my 'shade-tree programmer' skills but architecting the infrastructure has always been a bigger problem than I can solve on my own.   I really like how your module turned out.  It seems what I should focus on are creating robust functions to call and leave the rest of the logic to you.  I usually will test code on the hardware for a day or so looking for reliability.  

  I've already written the other function calls I mentioned but haven't yet tested them.  I'll put them in a branch on the original repo while I work on testing.  I haven't implemented checking the checksum byte yet but I probably should.  More advanced use is more likely to end up failing in some indeterminate way if the code isn't doing some level of input validation.

  The unit/measurement system is (I think) what Limor Fried talked about doing with a unified sensor library but never seemed to quite pull off, at least on python.  Compliments on the design!  And when you add in PID control services and the rest of what you've done with Mycodo - it creates a sensor/controller platform that can be built and deployed in a weekend.  Or less if your sensors are USB-attach.  The quality and performance of current sensors combined with Mycodo rival the performance of some pretty expensive handheld instrumentation, some that I own.  Mycodo is open-source at its best.

**Comment by kizniche** [comment]
133a2dcf is a nice example of the minimal amount of editing necessary to add a new sensor (if the units don't already exist).

Now that I have the new unit/measurement system working, I believe it's possible to contain all the necessary information about the sensor within the sensor module itself, allowing a single file to be dropped into Mycodo's ```input``` directory and for it to begin working immediately (assuming it's formatted correctly, of course).

This could allow for rapid testing of new input modules via an *input import* menu that can accept a file upload or URL. Does this sound like something that should be pursued?

**Comment by Theoi-Meteoroi** [comment]
From my perspective - that would be huge.   I have some favored sensors not yet supported (MCP9808, for instance) that I would probably work on to add.    Once you figure out how to get exactly one reading reliably - I can see how it could work well enough to support.  And of course that is always an important consideration - one that zombies or kills off most open-source projects.  I like fixing things and explaining how things work. Fortunately that turns out to be support in this context.  I think it is a great idea - and may activate some other contributors.

One interesting sensor I'd like to figure out how to integrate is the AMG88xx.  An example project with that device is here  [AMG88xx Explained](https://engineersportal.com/blog/2018/1/25/heat-mapping-with-a-64-pixel-infrared-sensor-and-raspberry-pi)  

Perhaps as a camera source selection?

**Comment by Theoi-Meteoroi** [comment]
I think I've completed a function library to manage and control the particle sensor.   I created a script that calls the functions in a loop so I can verify reasonable readings after re-enabling the sample fan.  Its been something of a discovery about how the data sheet is just a starting point and a somewhat poor translation.   They mention changing checksum schemes but don't explain further.  The return codes they provide for the fan enable command don't seem to work in practice. I find odd binary sequences on the read channel not described by the vendor immediately after a valid reading in streaming mode that I don't see when the device is in the Q&A mode.  

  I have a 4-channel Saleae logic analyzer that reveals what is really on the wire - and prevents certain premature aging on my part.  Between I2C, UART and I2S work, you can see what your code is really doing and the jerk devices it has to deal with.  The library test has run 650 loops with what look like continuous readings ( no weirdo outliers ) I think the library is good enough to call v1.0

  I don't know if this function library form is useful or easier to integrate with Mycodo.   Many sensors are only partially implemented and that always bugs me. I wanted to expose the entire interface with my implementation so the fan and power consumption could be controlled.  Older (cheaper) dust sensors depend on convection (with a heater element) and don't usually detect smaller than PM2.5.  It has a discrete inlet port that could be connected to a sample tube, if desired although I am testing the sensor out at the end of a ~2 meter USB cable so remote placement is not a problem either.

**Comment by Theoi-Meteoroi** [comment]
I think I found a description of the rest of those bytes [dfrobot device wiki](https://www.dfrobot.com/wiki/index.php/PM2.5_laser_dust_sensor_SKU:SEN0177#Communication_protocol)

Odd this wasn't in the vendor data sheet.

```
Communication protocol
Serial port baudrate: 9600; Parity: None; Stop Bits: 1; packet length is fixed at 32 bytes.


Start Character 1	0x42(fixed bit)
Start Character 2	0x4d(fixed bit)
Frame Length 16-byte	Frame Length = 2*9+2 (data+check bit)
Data 1, 16-byte	concentration of PM1.0, ug/m3
Data 2, 16-byte	concentration of PM2.5, ug/m3
Data 3, 16-byte	concentration of PM10.0, ug/m3
Data 4, 16-byte	Internal test data
Data 5, 16-byte	Internal test data
Data 6, 16-byte	Internal test data
Data 7, 16-byte	the number of particulate of diameter above 0.3um in 0.1 liters of air
Data 8, 16-byte	the number of particulate of diameter above 0.5um in 0.1 liters of air
Data 9, 16-byte	the number of particulate of diameter above 1.0um in 0.1 liters of air
Data 10, 16-byte	the number of particulate of diameter above 2.5um in 0.1 liters of air
Data 11, 16-byte	the number of particulate of diameter above 5.0um in 0.1 liters of air
Data 12, 16-byte	the number of particulate of diameter above 10.0um in 0.1 liters of air
Data 13, 16-byte	Internal test data
Check Bit for Data Sum, 16-byte	Check Bit = Start Character 1 + Start Character 2 + ...all data

```

**Comment by Theoi-Meteoroi** [comment]
I have found a fully realized ( complete with pull-ups and bus connections ) breakout board that will work on the hardware I2C or can also work with my GPIO I2C mux overlay.  Very simple to attach to a pi with the cable provided.  And really very inexpensive to add 8 I2C bus ports at US $6.90.  I tested the board with my overlay and it works fine.    You could use 5v sensors but would have to get the +5v from a source off the board since you have to power the board with 3.3v to be compatible with pi logic levels.  Check out the schematic for details.  

[DFRobot wiki for DFR0576](https://www.dfrobot.com/wiki/index.php/Gravity:_Digital_1-to-8_I2C_Multiplexer_SKU:DFR0576)

I don't see the board at any distributors now, so you probably have to order from China at the DFR website.  Mine took about two weeks to arrive.

**Comment by Theoi-Meteoroi** [comment]
Howdy,    I created a project on Hackaday.io for the GPIO mux device tree overlay with some additional background and hopefully a clear path to frustration-free installation.   I also found a vendor on TIndie that carries a small board that is like the DFR0576 but even less expensive at $5 a board.

[hackaday 9548a](https://hackaday.io/project/161718-pca9548a-i2c-mux-device-overlay-for-pi-gpio-i2c)  

[The Tindie source for 9548a mux boards](https://www.tindie.com/products/alexchu/zio-qwiic-mux-8-multiplexed-qwiic-i2c-busses/)
[cables for the boards](https://www.tindie.com/products/alexchu/qwiic-cable-10pcs-200mm/)

I've ordered some of these boards to test with Mycodo. Will report back results.

**Comment by kizniche** [comment]
That's great. Thanks for the info.
